fuzz/fuzz_pcap.c

```diff
@@ -0,0 +1,115 @@
+#include <config.h>
+#include "netdissect-stdinc.h"
+#include <sys/stat.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+
+#include <pcap/pcap.h>
+#include "netdissect.h"
+#include "print.h"
+
+FILE * outfile = NULL;
+
+static int
+fuzz_ndo_printf(netdissect_options *ndo, const char *fmt, ...)
+{
+    va_list args;
+    int ret;
+
+    va_start(args, fmt);
+    ret = vfprintf(outfile, fmt, args);
+    va_end(args);
+
+    return (ret);
+}
+
+void fuzz_openFile(const char * name) {
+    if (outfile != NULL) {
+        fclose(outfile);
+    }
+    outfile = fopen(name, "w");
+}
+
+static int bufferToFile(const char * name, const uint8_t *Data, size_t Size) {
+    FILE * fd;
+    if (remove(name) != 0) {
+        if (errno != ENOENT) {
+            printf("failed remove, errno=%d\n", errno);
+            return -1;
+        }
+    }
+    fd = fopen(name, "wb");
+    if (fd == NULL) {
+        printf("failed open, errno=%d\n", errno);
+        return -2;
+    }
+    if (fwrite (Data, 1, Size, fd) != Size) {
+        fclose(fd);
+        return -3;
+    }
+    fclose(fd);
+    return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+    pcap_t * pkts;
+    char errbuf[PCAP_ERRBUF_SIZE];
+    const u_char *pkt;
+    struct pcap_pkthdr *header;
+    int r;
+    u_int packets_captured = 0;
+    netdissect_options Ndo;
+
+    //initialize output file
+    if (outfile == NULL) {
+        outfile = fopen("/dev/null", "w");
+        if (outfile == NULL) {
+            return 0;
+        }
+    }
+
+    memset(&Ndo, 0, sizeof(Ndo));
+    ndo_set_function_pointers(&Ndo);
+    Ndo.program_name = "fuzz";
+    //avoid lookups
+    Ndo.ndo_nflag = 1;
+    //most verbose
+    Ndo.ndo_vflag = 5;
+    //to out outputfile
+    Ndo.ndo_printf=fuzz_ndo_printf;
+    init_print(&Ndo, 0, 0);
+
+    //rewrite buffer to a file as libpcap does not have buffer inputs
```

Comment:
> writing to a file and then reading it is a lot of overhead.
Indeed.

But as the comment said `libpcap does not have buffer inputs`
So the other solution would be to modify libpcap : have a new `pcap_open_buffer` interface and modify `pcap_open_file` so that both use the same internal logic.

For a first integration, I did not want to change the internal code.
That can be an optimization that comes afterwards.

Do you see any other way ?

Reply:
I have to take my comment about speed back. pcap_fuzz is about 10x as fast as bgp_print_fuzzer right now.  It's possible that this is due to vasprintf but as often happens, gut reactions about code speed can be wrong :-)

Reply:
I suspect oss-fuzz people use a ramdisk...
I am interested if you can get profiling data.
cf https://github.com/google/oss-fuzz/issues/2180#issuecomment-470886703